gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -392,23 +399,56 @@ private void addJars(Path jarFileDir, String jarFileList, Configuration conf)
     for (String jarFile : SPLITTER.split(jarFileList)) {
       Path srcJarFile = new Path(jarFile);
       FileStatus[] fileStatusList = lfs.globStatus(srcJarFile);
+
       for (FileStatus status : fileStatusList) {
-        // SNAPSHOT jars should not be shared, as different jobs may be using different versions of it
-        Path baseDir = status.getPath().getName().contains("SNAPSHOT") ? this.unsharedJarsDir : jarFileDir;
-        // DistributedCache requires absolute path, so we need to use makeQualified.
-        Path destJarFile = new Path(this.fs.makeQualified(baseDir), status.getPath().getName());
-        if (!this.fs.exists(destJarFile)) {
-          // Copy the jar file from local file system to HDFS
-          this.fs.copyFromLocalFile(status.getPath(), destJarFile);
+        // For each FileStatus there are chances it could fail in copying at the first attempt, due to file-existence
+        // or file-copy is ongoing by other job instance since all Gobblin jobs share the same jar file directory.
+        // the retryCount is to avoid cases (if any) where retry is going too far and causes job hanging.
+        int retryCount = 0 ;
+        boolean shouldFileBeAddedIntoDC = true;
+        Path destJarFile = calculateDestJarFile(status, jarFileDir);
+          // Adding destJarFile into HDFS until it exists and the size of file on targetPath matches the one on local path.
+        while(!this.fs.exists(destJarFile) || fs.getFileStatus(destJarFile).getLen() != status.getLen()) {
+          try {
+            if (fs.getFileStatus(destJarFile).getLen() != status.getLen()){
+              Thread.sleep(WAITING_TIME_ON_IMCOMPLETE_UPLOAD);
```

Comment:
15 seconds might not be sufficient time for a large jar. Can you make the retry parameter configurable so it can be bumped up if a job requires more time to copy jars?